File: akf_accounts/public/js/customizations/asset_capitalization_x.js
//  Mubashir Bashir 31-10-2025

function fetchWIPCompositeAssetPiItems(frm) {
    if (frm.doc.capitalization_method !== "Choose a WIP composite asset" || !frm.doc.target_asset) {
        console.log('Conditions not met for fetching WIP composite asset items');
        return;
    }

    // Show loading indicator
    frm.dashboard.set_headline_alert('Fetching WIP composite asset items from Purchase Invoice...');

    frm.clear_table('custom_purchase_invoice_items');

    frappe.call({
        method: 'akf_accounts.customizations.extends.asset_capitalization_x.get_wip_composite_asset_pi_items',
        args: {
            target_asset: frm.doc.target_asset
        },
        callback: function(response) {
            frm.dashboard.clear_headline();
            
            if (response.message && response.message.length > 0) {
                console.log('Found', response.message.length, 'matching Purchase Invoice Items from Purchase Invoice');
                
                let total_amount = 0;
                // Populate custom_purchase_invoice_items table
                response.message.forEach(function(pi_item) {
                    let child = frm.add_child('custom_purchase_invoice_items');

                    total_amount += flt(pi_item.amount);
                    
                    // Map basic fields
                    child.item = pi_item.item_code;
                    child.amount = pi_item.amount;
                    child.wip_account = pi_item.expense_account;
                    child.invoice_id = pi_item.invoice_id;
                    
                    // Map all accounting dimensions fields
                    const accountingDimensionFields = [
                        'donor_desk', 'fund_class', 'service_area', 'subservice_area',
                        'product', 'cost_center', 'transaction_type', 'inventory_scenario',
                        'project', 'task', 'asset_category', 'donor_type', 'donation_type',
                        'donor', 'reverse_donor', 'inventory_flag', 'rejected_inventory_flag',
                        'from_inventory_flag', 'from_inventory_donor', 'from_inventory_product',
                        'from_inventory_project', 'from_inventory_intention', 'from_inventory_donor_desk',
                        'from_inventory_donor_type', 'from_inventory_fund_class', 'from_inventory_cost_center',
                        'from_inventory_service_area', 'from_inventory_asset_category', 'from_inventory_subservice_area',
                        'from_inventory_transaction_type', 'from_product', 'from_service_area', 'from_subservice_area'
                    ];
                    
                    accountingDimensionFields.forEach(field => {
                        if (pi_item[field] !== undefined) {
                            child[field] = pi_item[field];
                        }
                    });
                });

                frm.set_value('custom_purchase_invoice_total', total_amount);

                frm.refresh_field('custom_purchase_invoice_items');
                
                frappe.show_alert({
                    message: __('Successfully fetched {0} items from submitted Purchase Invoices', [response.message.length]),
                    indicator: 'green'
                });
                
            } else {
                console.log('No matching Purchase Invoice Items found');
            }
        },
        error: function(err) {
            frm.dashboard.clear_headline();
            
            console.error('Error fetching WIP composite asset items from Purchase Invoice:', err);
            frappe.msgprint({
                title: __('Error'),
                message: __('Error fetching WIP composite asset items from Purchase Invoice. Please try again.'),
                indicator: 'red'
            });
        }
    });
}

function fetchWIPCompositeAssetSeItems(frm) {
    if (frm.doc.capitalization_method !== "Choose a WIP composite asset" || !frm.doc.target_asset) {
        console.log('Conditions not met for fetching WIP composite asset items');
        return;
    }

    // Show loading indicator
    frm.dashboard.set_headline_alert('Fetching WIP composite asset items from Stock Entry...');

    frm.clear_table('custom_stock_entry_items');

    frappe.call({
        method: 'akf_accounts.customizations.extends.asset_capitalization_x.get_wip_composite_asset_se_items',
        args: {
            target_asset: frm.doc.target_asset
        },
        callback: function(response) {
            frm.dashboard.clear_headline();
            
            if (response.message && response.message.length > 0) {
                console.log('Found', response.message.length, 'matching Stock Entry Items');
                
                let total_amount = 0;
                // Populate custom_purchase_invoice_items table
                response.message.forEach(function(se_item) {
                    let child = frm.add_child('custom_stock_entry_items');

                    total_amount += flt(se_item.amount);
                    
                    // Map basic fields
                    child.item = se_item.item_code;
                    child.amount = se_item.amount;
                    child.wip_account = se_item.expense_account;
                    child.invoice_id = se_item.invoice_id;
                    
                    // Map all accounting dimensions fields
                    const accountingDimensionFields = [
                        'donor_desk', 'fund_class', 'service_area', 'subservice_area',
                        'product', 'cost_center', 'transaction_type', 'inventory_scenario',
                        'project', 'task', 'asset_category', 'donor_type', 'donation_type',
                        'donor', 'reverse_donor', 'inventory_flag'
                    ];
                    
                    accountingDimensionFields.forEach(field => {
                        if (se_item[field] !== undefined) {
                            child[field] = se_item[field];
                        }
                    });
                });

                frm.set_value('custom_stock_entry_total', total_amount);

                frm.refresh_field('custom_stock_entry_items');
                
                frappe.show_alert({
                    message: __('Successfully fetched {0} items from submitted Stock Entry', [response.message.length]),
                    indicator: 'green'
                });
                
            } else {
                console.log('No matching Stock Entry Items found');
            }
        },
        error: function(err) {
            frm.dashboard.clear_headline();
            
            console.error('Error fetching WIP composite asset items from Stock Entry:', err);
            frappe.msgprint({
                title: __('Error'),
                message: __('Error fetching WIP composite asset items from Stock Entry. Please try again.'),
                indicator: 'red'
            });
        }
    });
}

frappe.ui.form.on('Asset Capitalization', {
    refresh: function(frm) {
        frm.add_custom_button(__('Fetch WIP Composite Items'), function() {
            fetchWIPCompositeAssetPiItems(frm);
            fetchWIPCompositeAssetSeItems(frm);
        }, __('Actions'));
    },
    
    capitalization_method: function(frm) {
        if (frm.doc.capitalization_method !== "Choose a WIP composite asset") {
            frm.clear_table('custom_purchase_invoice_items');
            frm.refresh_field('custom_purchase_invoice_items');
            frm.set_value('custom_purchase_invoice_total', 0);
            frm.refresh_field('custom_purchase_invoice_total');

            frm.clear_table('custom_stock_entry_items');
            frm.refresh_field('custom_stock_entry_items');
            frm.set_value('custom_stock_entry_total', 0);
            frm.refresh_field('custom_stock_entry_total');
        }
    },
    
    target_asset: function(frm) {
        if (!frm.doc.target_asset) {
            frm.clear_table('custom_purchase_invoice_items');
            frm.refresh_field('custom_purchase_invoice_items');
            frm.set_value('custom_purchase_invoice_total', 0);
            frm.refresh_field('custom_purchase_invoice_total');

            frm.clear_table('custom_stock_entry_items');
            frm.refresh_field('custom_stock_entry_items');
            frm.set_value('custom_stock_entry_total', 0);
            frm.refresh_field('custom_stock_entry_total');
        }
    }
});

File: akf_accounts/customizations/extends/asset_capitalization_x.py
import frappe
from frappe.utils import flt
from erpnext.assets.doctype.asset_capitalization.asset_capitalization import AssetCapitalization

class XAssetCapitalization(AssetCapitalization):
    def calculate_totals(self):
        super().calculate_totals()
        
        custom_pi_total = flt(self.custom_purchase_invoice_total or 0, self.precision("custom_purchase_invoice_total"))
        custom_se_total = flt(self.custom_stock_entry_total or 0, self.precision("custom_stock_entry_total"))
        
        self.total_value += custom_pi_total + custom_se_total
        self.total_value = flt(self.total_value, self.precision("total_value"))
        frappe.msgprint(f"Total Value after adding custom PI total: {self.total_value}")
        
        if self.target_qty:
            self.target_incoming_rate = self.total_value / self.target_qty

@frappe.whitelist()
def get_wip_composite_asset_pi_items(target_asset):
    """Get all Purchase Invoice Items for a WIP composite asset from submitted PIs"""
    items = frappe.db.sql("""
        SELECT 
            pi_item.item_code, 
            pi_item.amount, 
            pi_item.expense_account,
            pi_item.parent as invoice_id,
            pi_item.donor_desk, 
            pi_item.fund_class, 
            pi_item.service_area, 
            pi_item.subservice_area, 
            pi_item.product, 
            pi_item.cost_center,
            pi_item.transaction_type, 
            pi_item.inventory_scenario, 
            pi_item.project,
            pi_item.task, 
            pi_item.asset_category, 
            pi_item.donor_type, 
            pi_item.donation_type,
            pi_item.donor, 
            pi_item.reverse_donor, 
            pi_item.inventory_flag,
            pi_item.rejected_inventory_flag, 
            pi_item.from_inventory_flag,
            pi_item.from_inventory_donor,
            pi_item.from_inventory_product,
            pi_item.from_inventory_project,
            pi_item.from_inventory_intention,
            pi_item.from_inventory_donor_desk,
            pi_item.from_inventory_donor_type,
            pi_item.from_inventory_fund_class,
            pi_item.from_inventory_cost_center,
            pi_item.from_inventory_service_area,
            pi_item.from_inventory_asset_category,
            pi_item.from_inventory_subservice_area,
            pi_item.from_inventory_transaction_type,
            pi_item.from_product,
            pi_item.from_service_area,
            pi_item.from_subservice_area
        FROM `tabPurchase Invoice Item` pi_item
        INNER JOIN `tabPurchase Invoice` pi ON pi.name = pi_item.parent
        WHERE pi_item.wip_composite_asset = %s
        AND pi.docstatus = 1
        AND pi_item.parenttype = 'Purchase Invoice'
    """, target_asset, as_dict=1)
    
    return items


@frappe.whitelist()
def get_wip_composite_asset_se_items(target_asset):
    """Get all Stock Entry Items for a WIP composite asset from submitted SEs"""
    items = frappe.db.sql("""
        SELECT 
            se_item.item_code, 
            se_item.amount, 
            se_item.expense_account,
            se_item.parent as invoice_id,                          
            se_item.donor_desk, 
            se_item.fund_class, 
            se_item.service_area, 
            se_item.subservice_area, 
            se_item.product, 
            se_item.cost_center,
            se_item.transaction_type, 
            se_item.inventory_scenario, 
            se_item.project,
            se_item.task, 
            se_item.asset_category, 
            se_item.donor_type, 
            se_item.donation_type,
            se_item.donor, 
            se_item.reverse_donor, 
            se_item.inventory_flag
        FROM `tabStock Entry Detail` se_item
        INNER JOIN `tabStock Entry` se ON se.name = se_item.parent
        WHERE se_item.custom_wip_composite_asset = %s
        AND se.docstatus = 1
        AND se.stock_entry_type = 'Material Issue'
        AND se_item.parenttype = 'Stock Entry'
    """, target_asset, as_dict=1)
    
    return items

